### ISSUE TITLE
Bad assertion of array of objects

### ISSUE THREAD

**Issue by mitasov-ra** [lukas-krecan/JsonUnit#465]
**Describe the bug**
1. Describe which variant of the API are you using: AssertJ
2. Create a small reproducible example and paste it the the bug report.

This doesn't work:
```kotlin
assertThatJson("""
    [
      {"value": "1", "title": "Entity", "info": "Entity info"},
      {"value": "2", "title": "Column", "info": "Column info"},
      {"value": "3", "title": "Table", "info": "Table info"},
      {"value": "4", "title": "Schema", "info": "Schema info"}
    ]
""")
    .inPath("$[?(@.value =='1')]")
    .isArray.first()
    .isEqualTo(json("""{"value": "1", "title": "Entity", "info": "Entity info"}"""))
```

```log
[Different value found in node "$[?(@.value =='1')]" check first element] 
expected: {"value":"1","title":"Entity","info":"Entity info"}
 but was: {"info":"Entity info","title":"Entity","value":"1"}
```

But this does:
```kotlin
assertThatJson("""
    [
      {"value": "1", "title": "Entity", "info": "Entity info"},
      {"value": "2", "title": "Column", "info": "Column info"},
      {"value": "3", "title": "Table", "info": "Table info"},
      {"value": "4", "title": "Schema", "info": "Schema info"}
    ]
""")
    .inPath("$[?(@.value =='1')]")
    .node("[0]")
    .isEqualTo(json("""{"value": "1", "title": "Entity", "info": "Entity info"}"""))
```
3. Describe expected behavior

The above examples represent the same logic, and I expect them to work in the same way. It's a bit disappointing  that a more readable one doesn't work

**Comment by lukas-krecan** [comment]
Hi, thanks for reporting. Will try to fix that. The issue is caused by the fact that by calling `isArray()` you are moving from JsonUnit to AssertJ. And AssertJ does not know anything about Json comparison. I will take a look if there is a way to to change the behavior.

**Comment by lukas-krecan** [comment]
Latest findings. `first`, `last` and `element` methods are inherited from `ListAssert` and they return `ObjectAssert` which is not customizable. We can extend `FactoryBasedNavigableListAssert` and make those methods return `JsonMapAssert` but it's not a backward compatible change.

On the other hand it should break backward compatibility only for the methods that do not work anyway, so it may be possible to do. Have to investigate further.

**Comment by lukas-krecan** [comment]
Bad news. In order to change the assert, we would have to change the ELEMENT type in the `FactoryBasedNavigableListAssert` from Object to Map and that would break lot of stuff.

So the only way out of it is to mark those methods as deprecated and recommend using Json compatible alternatives.

**Comment by lukas-krecan** [comment]
And we can't even deprecate it since we are using the original ListAssert as return type of isArray.

**Comment by lukas-krecan** [comment]
Fixed in 2.29.0

**Comment by mitasov-ra** [comment]
@lukas-krecan I approve your decision to change return type of `isArray` to `JsonListAssert`, now `isEqualTo` works just fine.

But 2.30.0 version broke a lot of code, related to `isArray`, without giving the alternatives. For example, `JsonObjectAssert` does not provide such methods as `is<Something>` (`isObject`, `isArray`, etc.), `node`, `when` and so on.

I've come up with issue, that migrating the old code to the new API

```kt
assertThatJson("""[{"foo": "bar baz"}]""").isArray
.first()
.extracting("foo")
.asString()
.isEqualTo("bar baz")
```

seems impossible.

It'll be good to have something like 

```kt
assertThatJson("""[{"foo": "bar baz"}]""").isArray
.first()
.node("foo")
.isEqualTo(value("bar baz"))
```

or

```kt
assertThatJson("""[{"foo": "bar baz"}]""").isArray
.first()
.isObject
.contains(entry("foo", "bar baz")
```

But at the moment only `isEqualTo` is available, making such "check only one field" assertions hard to write and even harder to read.

P.S. btw thanks for your attention, it's really cool that you are so involved in this library
